### ISSUE TITLE
[Bug][XPU]: Silent persistent output corruption (endless "!" / token 0) under sustained concurrent decode on Arc Pro B70, W4A16 27B head_dim 256

### ISSUE THREAD

**Issue by bryanvine** [vllm-project/vllm#53480]
## Summary

A W4A16 (compressed-tensors, group 128, symmetric) 27B Qwen-family model served on a single Intel Arc Pro B70 (Battlemage, 32GB, xe driver) intermittently degenerates into emitting only "!" tokens (token id 0) under sustained concurrent decode. The corruption is silent: HTTP 200, finish_reason length, zero dmesg events, container healthy, /v1/models fine. It is PERSISTENT: after load drops, even fresh single-request generations return only "!" until the engine is restarted. Engine restart fully recovers (weight reload), which points at long-lived VRAM state (weights or a persistent workspace) being corrupted in place.

It struck our production engine four times in ~30 hours (three times at <= 4 concurrent requests after hours of sustained load; once within ~10 minutes when 12 concurrent synthetic streams mixed with real long-prompt traffic). Two sibling W4A16 models (14B and 8B, head_dim 128) on the same silicon under continuous scoring load have never corrupted in weeks, including c=12 stress sweeps.

The everything-token-0 signature suggests logits collapsing to NaN/zero (argmax degenerates to index 0), i.e. numeric poisoning of something persistent rather than per-request state.

## Environment

- GPU: Intel Arc Pro B70 32GB (Battlemage); reproduced on TWO different physical cards (VM, VFIO passthrough), so not a single-card hardware fault
- Host: kernel 6.17.0-41-generic, xe driver
- Image: intel/vllm:0.21.0-ubuntu24.04-20260805 (vLLM 0.21.1.dev18+g8df6feb7d)
  - torch 2.11.0+xpu, triton_xpu 3.7.0, vllm_xpu_kernels 0.1.8.2, dpcpp_cpp_rt 2025.3.2
- Model: Qwen3.8-27B fine-tune, W4A16 AutoRound -> compressed-tensors (group_size 128, symmetric, no g_idx), hidden 5120, 64 layers, head_dim 256, GQA 24/4, vocab 248320, lm_head unquantized
- Serve: --dtype bfloat16 --kv-cache-dtype auto --max-model-len 32768 --gpu-memory-utilization 0.97 --max-num-seqs 4..12 --max-num-batched-tokens 2048, no prefix caching
- Startup selects XPUwNa16LinearKernel (torch.ops._xpu_C.int4_gemm_w4a16) for CompressedTensorsWNA16; FLASH_ATTN (FA2) backend

## Fastest observed onset

Ramped concurrency sweep (cover-letter-shaped requests, max_tokens 700, 150s per level, c = 1, 2, 4, 8, 12) against the production engine while real traffic (6.7k-7.8k-token prompts, 2k-token completions) was also landing:

| c  | gen tok/s | mean e2e | garbage responses |
|----|-----------|----------|-------------------|
| 1  | 11.9      | 25.2s    | 0 |
| 2  | 26.0      | 24.4s    | 0 |
| 4  | 52.6      | 22.1s    | 0 |
| 8  | 102.6     | 23.3s    | 0 |
| 12 | 167.0     | 50.3s    | 36/36 (all "!") |

After the load stopped, a single-stream probe still returned 60 tokens of "!". Zero dmesg events throughout. Restart recovered.

## Isolation runs (all CLEAN, which is what makes this hard)

Each row is a separate fresh engine on the same hardware, 4-15 minute runs:

| model | image | gpu-mem-util | workload | result |
|-------|-------|--------------|----------|--------|
| 27B | dev18 | 0.90 | cold c=12, 133 reqs | clean |
| 27B | dev17 (intel/vllm:latest, 20260625 build) | 0.90 | cold c=12, 129 reqs | clean |
| 27B | dev17 | 0.97 | cold c=12, 134 reqs | clean |
| 27B | dev17 | 0.97 | full ramp c=1..12 | clean |
| 27B | dev18 | 0.97 | full ramp c=1..12 (exact replication of the corrupting run, minus live traffic) | clean |
| 27B | dev18 | 0.97 | 10 short-decode streams + 2 long-prompt (7k-token, chunked prefill) streams, 6 min | clean |
| 14B W4A16 head_dim 128 | dev18 | 0.5 | cold c=12, 239 reqs | clean |

So no single factor (image build, gpu-memory-utilization, ramp history, or a naive mixed prefill+decode workload) deterministically reproduces it from a cold engine in minutes. Production hit it four times in 30 hours. It behaves like a race whose probability rises steeply with decode batch depth and with real-traffic batch compositions.

## What we ruled out

- Hardware: identical corruption on two different physical cards; a c=12 sweep on the second card corrupted the same way while weeks of equal-wattage load from the head_dim-128 models never did. Power delta between clean c=8 (195W) and corrupting c=12 (199W) is ~2%.
- The int4 GEMM kernel in isolation: a standalone harness driving torch.ops._xpu_C.int4_gemm_w4a16 with this model's exact layer shapes (5120x8192, 6144x5120, 5120x34816, 17408x5120, group 128) at M = 1..32, 64, 128, 512, 2048 shows no NaN/Inf, bit-exact determinism across repeated calls, no input-buffer mutation (weight/scale checksums stable), and no cross-call poisoning of a baseline M=1 result.
- vLLM Python-level differences: the dev17 and dev18 images' /opt/vllm/vllm trees are byte-identical except _version.py and an API-middleware helper; torch/triton/vllm_xpu_kernels/dpcpp versions are identical. (We initially suspected a build regression; the tree diff killed that.)
- KV preemption: vLLM logged zero preemptions in all runs (KV was never near full; 172k-token cache vs ~10-40k in flight).

## Notes that may help localize it

- head_dim 256 + GQA 24/4 is the standout difference from our never-corrupting 14B/8B (head_dim 128) neighbors; vllm_xpu_kernels 0.1.8.2 predates the v0.1.9 mixed-batch attention rework.
- vocab 248320 means the (unquantized bf16) lm_head GEMM's M equals the decode batch depth; corruption onset correlates with deeper decode batches.
- The corruption survives across requests but not restarts: whatever is poisoned lives in VRAM for the process lifetime (weights, persistent workspace, or compiled-graph buffers).
- Ubuntu 24.04 HWE-class host kernel with xe; the same box previously produced GP-fault/BCS-reset issues under TP=2 (#41663) but this failure mode has no driver-visible events at all.

## Ask

We are happy to run instrumented builds, extra logging, or targeted experiments on this hardware; the model and full configs can be shared. Detection guidance for other users: only a CONTENT probe catches this (health endpoints and /v1/models keep answering normally); we run a cron that generates a short completion every 10 minutes and restarts on "!" output.

Mitigation we are validating: moving the affected engine to the 20260625 build at gpu-memory-utilization 0.90 (every isolation cell on that config was clean, and it is ~20% faster on this model: 14.8 vs 11.9 tok/s single-stream, 154.9 tok/s at c=12).

**Comment by bryanvine** [comment]
Recurrence on the mitigation posture, 2026-08-26 17:20:04 UTC.

Since 2026-08-23 ~22:00 UTC this engine has been running what I described above as the mitigation: stock `intel/vllm:latest` (0.21.1.dev17+g0a475), `--gpu-memory-utilization 0.90`, no patched-kernel bind mounts. It ran clean for roughly three days and then degenerated again in exactly the same way: a content canary that sends a fixed short prompt every 20 minutes got a reply consisting only of `!` repeated to the token cap, with `finish_reason` normal. Container health and `/v1/models` stayed green throughout, as before. The canary restarted the engine and it has been clean for the ~14h since.

The one variable that changed between the clean stretch and the recurrence: on 2026-08-24 I raised `--max-num-seqs` from 4 to 12 (and the matching gateway concurrency cap). So the corruption reproduces on an unpatched upstream build, and raising sustained concurrent decode width from 4 to 12 is the only deliberate change preceding it. That is consistent with the concurrency-dependent race hypothesis rather than anything in the quantized weights or the custom kernels, both of which were exonerated in the isolation matrix in the original report.

Posture at the time of the recurrence:

```
image                     intel/vllm:latest (0.21.1.dev17+g0a475)
--gpu-memory-utilization  0.90
--max-num-seqs            12
patched kernel mounts     none
model                     Qwen3.8-27B W4A16 (head_dim 256), single card, no TP
card                      Arc Pro B70, guest card0 (07:00.0), PCIe passthrough to a VM
```

No driver events, no devcoredump, and no kernel log entries around the failure window; the card kept serving requests, it just served corrupted tokens. Detection still requires a content probe, since every liveness surface reports healthy.

Happy to run a specific instrumented build or dump state at the next occurrence if that would help narrow it.

**Comment by bryanvine** [comment]
Two more recurrences on the mitigation posture since my 2026-08-26 report, and the interval is shortening:

- 2026-08-26 17:20:04 UTC (already reported above)
- 2026-08-27 19:10:04 UTC
- 2026-08-28 02:20:04 UTC

Same signature each time: the served model starts emitting an unbroken run of `!` for every request while the container stays healthy and `/v1/models` keeps answering 200. A 10-minute content canary catches it and restarts the engine; each restart comes back clean and serves normally (current uptime 5h, `finish_reason: stop`, coherent output on probe).

Unchanged environment across all three: `intel/vllm:latest` (0.21.1.dev17+g0a475), gpu-memory-utilization 0.90, max-num-seqs 4, no custom kernel patches mounted, Qwen3.8-27B W4A16 on a single Arc card (07:00.0) inside a VM with PCIe passthrough. No driver events, no devcoredump, nothing in dmesg on the guest around any of the three timestamps.

Gap between events: ~72h, then ~26h, then ~7h. Load was steady across the window (roughly 300 generation requests/day), so this looks like elapsed-time or cumulative-request drift rather than a load spike. Happy to collect any specific engine state before the next restart if there is something worth dumping.

**Comment by bryanvine** [comment]
Correction to my comment above: I quoted `max-num-seqs 4`, but the running engine is `--max-num-seqs 12`. It was raised from 4 to 12 on 2026-08-24, after the mitigation cutover, and the gateway's parallelism cap was raised to match. Everything else in that comment stands (`intel/vllm:latest` 0.21.1.dev17+g0a475, `--gpu-memory-utilization 0.90`, `--max-model-len 32768`, no custom kernel patches, Qwen3.8-27B W4A16 on one Arc card via VM passthrough).

That makes the timeline more suggestive than I implied:

- 2026-08-23 22:00 UTC: mitigation deployed, `max-num-seqs 4`
- 2026-08-24: raised to 12
- 2026-08-26 17:20, 08-27 19:10, 08-28 02:20 UTC: corruption, gaps ~72h, ~26h, ~7h

So the only fully clean stretch overlaps the period at width 4, and every recurrence has been at width 12. That is one variable change and three data points, not a controlled result, but concurrency width is now our leading suspect rather than elapsed time. If it holds up I will drop back to 4 and report whether the interval stretches again.

**Comment by bryanvine** [comment]
One further recurrence on the mitigation posture, after the longest clean gap so far.

- 2026-08-26 17:20:04 UTC
- 2026-08-27 19:10:04 UTC
- 2026-08-28 02:20:04 UTC
- **2026-09-02 02:30:04 UTC** (new)

So the shortening interval I flagged on 2026-08-28 did not continue: 2026-08-29 through 2026-09-01 were clean (roughly 4.5 days, ~650 canary probes at a 10-minute cadence), then the same failure returned. Signature is unchanged, an unbroken run of `!` to the token cap, `finish_reason` length, container health and `/v1/models` both reporting fine throughout.

Engine is still the posture described earlier in the thread: `intel/vllm:latest` (dev17, `vllm-0.21.1.dev17+g0a4756bb5`), `--gpu-memory-utilization 0.90`, no out-of-tree patches, `--max-num-seqs 12`, single card, W4A16 Qwen3.8-27B on Intel XPU. No driver events in dmesg around the failure, and no devcoredump was produced.

The intermittency profile now looks like a low-rate race rather than anything that correlates with load or uptime: the engine had been up and serving continuously across the clean stretch, and the failure landed in a low-traffic hour. Adding a second data point for anyone trying to reproduce, elapsed uptime before the event does not appear to be the variable.

Content-level probing remains the only thing that catches it. Happy to run a specific instrumented build if that would help narrow it.

**Comment by bryanvine** [comment]
Same-day second recurrence, which breaks the "low-rate" reading from my last comment.

- 2026-08-26 17:20:04 UTC
- 2026-08-27 19:10:04 UTC
- 2026-08-28 02:20:04 UTC
- 2026-09-02 02:30:04 UTC
- **2026-09-02 14:50:04 UTC** (new)

The gap between the last two is 12h20m, against the 4.5-day clean stretch that immediately preceded them. So the failure is not spread evenly in time; it clusters. Both 09-02 events hit the same engine process generation only in the sense that the 02:30 restart brought up a fresh engine, which then failed again 12 hours later. Signature identical to every prior instance: unbroken `!` to the token cap, `finish_reason` length, container health and `/v1/models` clean throughout, no dmesg driver events, no devcoredump.

Posture unchanged: `intel/vllm:latest` (`vllm-0.21.1.dev17+g0a4756bb5`), `--gpu-memory-utilization 0.90`, no out-of-tree patches, `--max-num-seqs 12`, single card, W4A16 Qwen3.8-27B on Intel XPU.

One new isolation data point. Since 2026-09-02 01:58 UTC I have run an equivalent content canary against a **second, independent engine on the other card of the same box**: same `dev17` image, but a different model and shape (W4A16 Qwen3-14B, `--max-num-seqs 12`, scoring-shaped prompts, 60 token generations, 10-minute cadence offset from the writer probe). That engine has been clean for the whole window, roughly 175 probes, including across both writer failures. Earlier in this thread I reported the degeneration on both cards, so this is not evidence that the second card is immune; it is evidence that the two engines do not fail together, which argues against anything box-wide (driver state, shared allocator pressure, thermal) and for something local to an engine process.

The other difference between the two engines is generation length: the writer produces 500 to 2000 token documents, the scorer produces about 20 tokens of JSON. If the race needs a long decode to expose itself, that would fit both the clustering and the card-symmetry I saw before.

Content-level probing is still the only thing that detects this. Still happy to run an instrumented build.

**Comment by bryanvine** [comment]
Another recurrence on the mitigation posture, after the longest clean gap yet.

- 2026-08-26 17:20:04 UTC
- 2026-08-27 19:10:04 UTC
- 2026-08-28 02:20:04 UTC
- 2026-09-02 02:30:04 UTC
- 2026-09-02 14:50:04 UTC
- **2026-09-12 07:40:04 UTC** (new)

The gap since the previous event is 9 days 16h 50m, roughly double the 4.5-day stretch before the 09-02 pair. Signature identical to every prior instance: unbroken `!` to the token cap (one production call at 07:37:27 UTC ran to the 6144-token limit about three minutes before the canary probe caught it), container health and `/v1/models` clean, no dmesg driver events, no devcoredump. A plain `docker restart` recovered it (engine back at 07:40:08 UTC, `Application startup complete` shortly after), and it has been clean for the 24 hours since.

Posture unchanged: `intel/vllm:latest` (`vllm-0.21.1.dev17+g0a4756bb5`), `--gpu-memory-utilization 0.90`, no out-of-tree patches, `--max-num-seqs 12`, single card, W4A16 Qwen3.8-27B on Intel XPU.

The independent scoring engine on the other card of the same box (same `dev17` image, W4A16 Qwen3-14B, short JSON generations, 10-minute content canary) logged no garbage across this event either. That makes three writer failures in a row with no matching failure on the short-decode engine, which still fits the reading that the race needs long decodes to show up and is local to an engine process rather than box-wide.

**Comment by faaany** [comment]
@bryanvine thanks for raising this issue. I will investigate this on my end.

**Comment by faaany** [comment]
@bryanvine your model (Qwen3.8-27B) is a hybrid architecture: 48 of the 64 layers are GDN linear attention. That is the real difference from your pure-attention 14B/8B. This PR on vllm-xpu-kernels (https://github.com/vllm-project/vllm-xpu-kernels/pull/411) fixed a shared-local-memory race issue in the GDN chunk kernels that produces NaN logits ('!!!') on Arc Pro B70. The fix landed in v0.1.10, so it's not in the vllm_xpu_kernels 0.1.8.2 your image ships.

Could you pls try the latest docker image `docker pull vllm/vllm-openai-xpu:v0.29.0` and see whether your case still persists?  If it still reproduces, please let us know.

**Comment by bryanvine** [comment]
@faaany thank you, that explains the 27B-only pattern. Tested today on the same Arc Pro B70 box:

- Image: `vllm/vllm-openai-xpu:v0.29.0` (vllm 0.29.0+xpu, vllm_xpu_kernels 0.1.14.1, torch 2.13.0+xpu), replacing `intel/vllm:latest` (0.21.1.dev17, vllm_xpu_kernels 0.1.8.2).
- Model: the same Qwen3.8-27B W4A16 checkpoint, `--max-num-seqs 12 --gpu-memory-utilization 0.90 --max-model-len 32768`, single card.
- Repro condition: 12 parallel streams of ~700-token generations in 150 s rounds. On the old image this produced the `!!!!` degeneration within one round (36/36 responses garbage, corruption persisting after load dropped).
- Result on v0.29.0: 4 rounds, 318 requests, 0 garbage, 0 errors, single-stream probes clean afterwards, no engine resets in dmesg. Aggregate throughput 135 to 144 tok/s at c=12 (about 10% under the old image's 155).

The writer is now serving production traffic on v0.29.0 with the same content canary that caught the earlier recurrences (last one on the old image: 2026-09-12). I will report back here if it reproduces again; otherwise consider the GDN chunk kernel fix confirmed for this case.
